lib/dns_over_udp.toit

```diff
@@ -0,0 +1,277 @@
+// Copyright (C) 2020 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the lib/LICENSE file.
+
+import binary show BIG_ENDIAN
+import net
+
+DNS_DEFAULT_TIMEOUT ::= Duration --s=20
+DNS_RETRY_TIMEOUT ::= Duration --s=1
+HOSTS_ ::= {"localhost": "127.0.0.1"}
+CACHE_ ::= Map  // From name to CacheEntry.
+MAX_CACHE_SIZE_ ::= platform == "FreeRTOS" ? 30 : 1000
+MAX_TRIMMED_CACHE_SIZE_ ::= MAX_CACHE_SIZE_ / 3 * 2
+
+class DnsException:
+  text/string
+
+  constructor .text:
+
+  stringify -> string:
+    return "DNS lookup exception $text"
+
+ip_string_ name/string -> bool:
+  dots := 0
+  name.do:
+    if it == '.':
+      dots++
+    else:
+      if not '0' <= it <= '9': return false
+  return dots == 3
+
+/**
+Look up a domain name.
+
+If given a numeric address like 127.0.0.1 it merely parses
+  the numbers without a network round trip.
+
+Does not currently cache results, so there is normally a
+  network round trip on every use.
+
+By default the server is "8.8.8.8" which is the Google DNS
+  service.
+
+Currently only works for IPv4, not IPv6.
+*/
+lookup -> net.IpAddress
+    host/string
+    [get_udp_socket]
+    --server/string="8.8.8.8"
+    --timeout/Duration=DNS_DEFAULT_TIMEOUT:
+  if ip_string_ host:
+    return net.IpAddress.parse host
+  q := DnsQuery_ host
+  return q.get --server=server --timeout=timeout get_udp_socket
+
+class DnsQuery_:
+  static A     ::= 1  // A address.
+  static CNAME ::= 5  // Canonical name.
+  static CLASS ::= 1  // The Internet class.
+  id/int
+  name/string
+
+  constructor .name:
+    id = random 0x10000
+
+  /**
+  Look up a domain name.
+
+  If given a numeric address like "127.0.0.1" it merely parses
+    the numbers without a network round trip.
+
+  By default the server is "8.8.8.8" which is the Google DNS
+    service.
+
+  Currently only works for IPv4, not IPv6.
+  */
+  get -> net.IpAddress
+      --server/string="8.8.8.8"
+      --timeout/Duration=DNS_DEFAULT_TIMEOUT
+      [get_udp_socket]:
+    if HOSTS_.contains name:
+      return net.IpAddress.parse HOSTS_[name]
+
+    hit := find_in_cache_ server
+    if hit: return hit
+
+    query := create_query_
+    socket := get_udp_socket.call
```

Comment:
Maybe add a type to have a dynamic check that it is the correct socket.
```suggestion
    socket/udp.Socket := get_udp_socket.call
```
You would also need to have the import:
```
import net.udp
```
Finally, we should probably mention it in the toit doc.

Reply:
I looked at the previous implementation and it should create a socket of the correct type.